Error Handling Section (#78)
chhoumann/p4: Report/sections/3_implementation/error_handling.tex

\section{Error handling}
It is useful for users to get error-messages detailing any syntax errors there might have been found during parsing.

> A simple solution is to provide the user with the offending token and/or line. This is very basic error output, which could be augmented by a detailed message telling the user what is wrong; i.e. a missing semicolon. Even better if the message told the user how to fix the error. 

Reviewer: ```suggestion
A simple solution is to provide the user with the offending token and/or line. This is very basic error output, which could be augmented by a detailed message telling the user what is wrong. A missing semicolon, for example. Even better if the message told the user how to fix the error. 
```

> Going a level further, the parser could attempt to make a recovery. Handle the error, however possible it may be. There are a few ways to do this.

Reviewer: ```suggestion
Going a level further, the parser could attempt to make a recovery by handling the error, if possible. There are a few ways to do this.
```

First is panic-mode recovery, in which the parser removes input tokens until a synchronizing token is found. In the case of \dazel{}, this could be a semicolon, indicating the end of a statement. This is a simple solution, but it may skip a considerable amount of input, which may produce unexpected results\cite{crafting_a_compiler}.

> It is also possible to do phrase-level recovery. This means to perform corrections locally in some erroneous phrase. It could be to insert a missing semicolon, or even delete an extraneous one. In performing corrections, one must be careful not to enter an infinite loop\cite{spo_course}.

Reviewer 2: ```suggestion
It is also possible to do phrase-level recovery. This means performing corrections locally in some erroneous phrase, for example inserting a missing semicolon or even deleting an extraneous one. In performing corrections, one must be careful not to enter an infinite loop\cite{spo_course}.
```

Author: ```suggestion
It is also possible to do phrase-level recovery. This means performing corrections locally in some erroneous phrase. This could be something as trivial as inserting a missing semicolon, or even deleting an extraneous one. In performing corrections, one must be careful not to enter an infinite loop\cite{spo_course}.
```

The \texttt{LogMessageReceived} event, seen on line \ref{code:LogMessageReceived}, is used by a custom-built console, shown in the \dazel{} game. An example can be seen in \figref{dazelconsole}, showing both warnings and messages. The messages shown in this particular example represent usages of \dazel{}'s \texttt{Print} function. In another example, seen in \figref{dazelconsoleerror}, an error is displayed. This error is due to writing "this is an error" in a \texttt{GameObjectContent} block. Lastly, \figref{fig:dazel_console_custom_error} shows the \dazel{} console with a type error.

\fig{1}{dazel_console}{The \dazel{} console with no error.\label{fig:dazelconsole}}
\fig{1}{dazel_console_error}{The \dazel{} console with a syntax error.\label{fig:dazelconsoleerror}}
\fig{1}{dazel_console_custom_error}{The \dazel{} console with a type error.\label{fig:dazel_console_custom_error}}

Having proper error output in a compiler is, as previously mentioned, very important. \dazel{} has detection of errors in multiple phases, which allows it to give a more detailed and descriptive error message. A clear next step would be to implement error recovery, as described earlier in this section.

> The next chapter details how we've ensured that \dazel{} runs as it should - through testing.

Reviewer: ```suggestion
The next chapter details how we have ensured that \dazel{} runs as it should - through testing.
```

\begin{lstlisting}[language=CSharp, caption={The altered \texttt{SyntaxError} method.}, label={lst:SyntaxErrorCode}, escapechar=|]
public override void SyntaxError(TextWriter output, IRecognizer recognizer, 
    IToken offendingSymbol, int line, int charPositionInLine,
    string msg, RecognitionException e)
{
    StringBuilder errorLog = new StringBuilder();
    base.SyntaxError(output, recognizer, offendingSymbol, line, charPositionInLine, msg, e);
    
    // UnderlineError(...) will not be shown in the spirit of brevity.
    string underlineError = UnderlineError(recognizer, offendingSymbol, line, charPositionInLine);

    errorLog.AppendLine($"File: {recognizer.InputStream.SourceName}");
    errorLog.AppendLine($"Line {line}:{charPositionInLine} at {offendingSymbol}: {msg}\n");
    errorLog.AppendLine($"Error: \n{underlineError}");

    DazelLogger.EmitError(errorLog.ToString(), offendingSymbol);|\label{code:dazellogger}|
}
\end{lstlisting}

This leads us to the second interface. \dazel{} uses a custom-built error logging system, which can be seen invoked on line \ref{code:dazellogger}.

> The \texttt{DazelLogger} is a static class which allows the \dazel{} compiler to log errors. There are three methods, \texttt{EmitError}, \texttt{EmitWarning}, and \texttt{EmitMessage}. These represent the levels of messages that the user may be presented with. Every member of this class is \texttt{static}, such that they can be accessed from anywhere, without instantiating a new logger - and without unnecessary dependency injection.

Reviewer 2: ```suggestion
The \texttt{DazelLogger} is a static class which allows the \dazel{} compiler to log errors. The class contains three methods, \texttt{EmitError}, \texttt{EmitWarning}, and \texttt{EmitMessage}. These represent the levels of messages that the user may be presented with. Every member of this class is \texttt{static}, such that they can be accessed from anywhere, without instantiating a new logger - and without unnecessary dependency injection.
```

The most interesting case is \texttt{EmitError}, which can be seen in code \snipref{lst:EmitErrorCode}. It sets the \texttt{HasErrors} flag to true, such that the \dazel{} compiler can react accordingly. Likewise, it formats the output message and invokes an event, passing it in as a parameter.

> If the \texttt{ThrowExceptions} flag is set to true, the function will also throw an error.

Reviewer 2: ```suggestion
If the \texttt{ThrowExceptions} flag is set to true, the function will also throw an exception inside the Unity log. Note that this flag can be set inside of Unity to aid in debugging the \dazel{} compiler - it is not something the user of can access.
```

Author: ```suggestion
If the \texttt{ThrowExceptions} flag is set to true, the function will also throw an exception.
```

\begin{lstlisting}[language=CSharp, caption={\texttt{EmitError} in the \texttt{DazelLogger}.}, label={lst:EmitErrorCode}, escapechar=|]
public static void EmitError(string message, IToken token)
{
    HasErrors = true;
    
    string output = $"Error on line {token.Line} in {token.InputStream.SourceName}:\n{message}";
    LogMessageReceived?.Invoke(output, LogType.Error);|\label{code:LogMessageReceived}|

    if (ThrowExceptions)
    {
        throw new Exception(message);
    }
}
\end{lstlisting}

> The \texttt{LogMessageReceived} event, seen on line \ref{code:LogMessageReceived}, is used by a custom-built console, shown in the \dazel{} game. An example can be seen in \figref{dazelconsole}, showing both warnings and messages. The messages shown in this particular example represent usages of \dazel{}'s \texttt{Print} function. In another example, seen in \figref{dazelconsoleerror}, an error is displayed. This error is due to writing "this is an error" in a \texttt{GameObjectContent} block. Lastly, \figref{fig:dazel_console_custom_error} shows the \dazel{} console with a type error.

Reviewer 2: ```suggestion
The \texttt{LogMessageReceived} event, seen on line \ref{code:LogMessageReceived}, is used by a console GUI in the \dazel{} game. An example can be seen in \figref{dazelconsole}, showing both warnings and messages. The messages shown in this particular example represent usages of \dazel{}'s \texttt{Print} function. In another example, seen in \figref{dazelconsoleerror}, an error is displayed. This error is due to writing "this is an error" in a \texttt{GameObjectContent} block, which is picked up by the parser. Lastly, \figref{dazel_console_custom_error} shows the \dazel{} console with a type error.
```
